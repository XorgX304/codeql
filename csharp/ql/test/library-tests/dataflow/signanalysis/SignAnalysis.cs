using System;
using System.Linq;
using System.Diagnostics;

class SignAnalysis
{
    static int GetRandomValue() { return (new System.Random()).Next(0, 1000) - 500; }

    int RandomValue { get => GetRandomValue(); }

    int random = GetRandomValue();

    int SsaSources(int p, int[] values)
    {
        var v = GetRandomValue();
        if (v < 0)
        {
            return v;
        }

        v = RandomValue;
        if (v < 0)
        {
            return v;
        }

        v = p;
        if (v < 0)
        {
            return v;
        }

        v = random;
        if (v < 0)
        {
            return v;
        }

        v = values[0];
        if (v < 0)
        {
            return v;
        }

        int x = values[1];
        v = x;
        if (v < 0)
        {
            return v;
        }

        return 0;
    }

    void Operations(int i, int j, bool b)
    {
        if (i < 0 && j < 0)
        {
            var x = i + j;
            System.Console.WriteLine(x); // strictly neg
            x = i * j;
            System.Console.WriteLine(x); // strictly pos
            x = i / j;
            System.Console.WriteLine(x); // pos
            x = i - j;
            System.Console.WriteLine(x); // no clue
            x = i % j;
            System.Console.WriteLine(x); // neg
            x = i++;
            System.Console.WriteLine(x); // strictly neg
            x = i--;
            System.Console.WriteLine(x); // neg
            x = -i;
            System.Console.WriteLine(x); // strictly pos
            x = +i;
            System.Console.WriteLine(x); // strictly neg
            var l = (long)i;
            System.Console.WriteLine(l); // strictly neg

            x = i;
            x += i;
            System.Console.WriteLine(x); // strictly neg
        }

        if (i < 0 && j > 0)
        {
            var x = i + j;
            System.Console.WriteLine(x);
            x = i * j;
            System.Console.WriteLine(x); // strictly neg
            x = i / j;
            System.Console.WriteLine(x); // neg
            x = i - j;
            System.Console.WriteLine(x); // strictly neg
            x = i % j;
            System.Console.WriteLine(x); // neg
            x = b ? i : j;
            System.Console.WriteLine(x); // any (except 0)
        }
    }

    void NumericalTypes()
    {
        var f = 4.2f;
        System.Console.WriteLine(f);
        var d = 4.2;
        System.Console.WriteLine(d);
        var de = 4.2m;
        System.Console.WriteLine(de);
        var c = 'a';
        System.Console.WriteLine(c);
    }

    int f0;

    int f1;

    void Field0()
    {
        f0++;
        System.Console.WriteLine(f0); // strictly positive
        f0 = 0;
    }

    void Field1()
    {
        f1++;
        System.Console.WriteLine(f1); // no clue
        f1 = -10;
    }

    void Field2()
    {
        System.Console.WriteLine(f1); // no clue
    }

    void Ctor()
    {
        var i = new Int32(); // const 0 value
        i++;
        System.Console.WriteLine(i); // strictly pos
    }

    int Guards(int x, int y)
    {
        if (x < 0)
        {
            return x; // strictly negative
        }

        if (y == 1)
        {
            return y; // strictly positive
        }

        if (y is -1)
        {
            return y; // strictly negative [MISSING]
        }

        if (x < y)
        {
            return y; // strictly positive
        }

        var b = y == 1;
        if (b)
        {
            return y; // strictly positive
        }

        return 0;
    }

    void Inconsistent()
    {
        var i = 1;
        if (i < 0)
        {
            System.Console.WriteLine(i); // reported as strictly pos, although unreachable
        }
    }

    void SpecialValues(int[] ints)
    {
        System.Console.WriteLine(ints.Length); // positive
        ints = new int[] { 1, 2, 3 };
        System.Console.WriteLine(ints.Length); // 3, so strictly positive
        System.Console.WriteLine(ints.Count()); // positive
        System.Console.WriteLine(ints.Count(i => i > 1)); // positive

        var s = "abc";
        System.Console.WriteLine(s.Length); // positive, could be strictly positive

        var enumerable = Enumerable.Empty<int>();
        System.Console.WriteLine(enumerable.Count()); // positive

        var i = new int[,] { { 1, 1 }, { 1, 2 }, { 1, 3 } };
        System.Console.WriteLine(i.Length); // 6, so strictly positive
    }

    void Phi1(int i)
    {
        if (i > 0)
        {
            System.Console.WriteLine(i); // strictly positive
        }
        else
        {
            System.Console.WriteLine(i); // negative
        }
        System.Console.WriteLine(i); // any
    }

    void Phi2(int i)
    {
        if (i > 0)
        {
            System.Console.WriteLine(i); // strictly positive
        }
        else
        {
            if (i < 0) // negative
            {
                System.Console.WriteLine(i); // strictly negative
                return;
            }
        }
        System.Console.WriteLine(i); // positive, not found
    }

    void Phi3(int i)
    {
        if (i > 0)
        {
            System.Console.WriteLine(i); // strictly positive
        }
        else
        {
            if (i < 0) // negative
            {
                System.Console.WriteLine(i); // strictly negative
            }
            else
            {
                System.Console.WriteLine(i); // zero, nothing is reported
            }
        }
    }

    void Loop(int i, int j, int k)
    {
        if (i > 0)
        {
            while (i >= 0) // any
            {
                i--; // positive
                System.Console.WriteLine(i); // any
            }
            System.Console.WriteLine(i); // strictly neg
        }

        if (j > 0)
        {
            while (j > 0)
            {
                j--; // strictly pos
                System.Console.WriteLine(j); // positive
            }
            System.Console.WriteLine(j); // reported negative, can only be 0
        }

        if (k > 0)
        {
            while (k > 0)
            {
                k--; // strictly pos
                System.Console.WriteLine(k); // positive

                if (k == 5) // positive
                {
                    break;
                }
            }
            System.Console.WriteLine(k); // any
        }
    }

    void Assert(int i, bool b)
    {
        Debug.Assert(i > 0);
        System.Console.WriteLine(i); // strictly positive

        if (b)
            System.Console.WriteLine(i); // strictly positive
    }
}

// semmle-extractor-options: /r:System.Linq.dll