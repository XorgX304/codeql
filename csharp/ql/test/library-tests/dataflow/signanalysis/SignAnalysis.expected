| SignAnalysis.cs:7:72:7:75 | 1000 | positive strictlyPositive |
| SignAnalysis.cs:7:80:7:82 | 500 | positive strictlyPositive |
| SignAnalysis.cs:18:20:18:20 | access to local variable v | negative strictlyNegative |
| SignAnalysis.cs:24:20:24:20 | access to local variable v | negative strictlyNegative |
| SignAnalysis.cs:30:20:30:20 | access to local variable v | negative strictlyNegative |
| SignAnalysis.cs:36:20:36:20 | access to local variable v | negative strictlyNegative |
| SignAnalysis.cs:42:20:42:20 | access to local variable v | negative strictlyNegative |
| SignAnalysis.cs:45:24:45:24 | 1 | positive strictlyPositive |
| SignAnalysis.cs:49:20:49:20 | access to local variable v | negative strictlyNegative |
| SignAnalysis.cs:59:21:59:21 | access to parameter i | negative strictlyNegative |
| SignAnalysis.cs:59:21:59:25 | ... + ... | negative strictlyNegative |
| SignAnalysis.cs:59:25:59:25 | access to parameter j | negative strictlyNegative |
| SignAnalysis.cs:60:38:60:38 | access to local variable x | negative strictlyNegative |
| SignAnalysis.cs:61:13:61:21 | ... = ... | positive strictlyPositive |
| SignAnalysis.cs:61:17:61:17 | access to parameter i | negative strictlyNegative |
| SignAnalysis.cs:61:17:61:21 | ... * ... | positive strictlyPositive |
| SignAnalysis.cs:61:21:61:21 | access to parameter j | negative strictlyNegative |
| SignAnalysis.cs:62:38:62:38 | access to local variable x | positive strictlyPositive |
| SignAnalysis.cs:63:13:63:21 | ... = ... | positive |
| SignAnalysis.cs:63:17:63:17 | access to parameter i | negative strictlyNegative |
| SignAnalysis.cs:63:17:63:21 | ... / ... | positive |
| SignAnalysis.cs:63:21:63:21 | access to parameter j | negative strictlyNegative |
| SignAnalysis.cs:64:38:64:38 | access to local variable x | positive |
| SignAnalysis.cs:65:17:65:17 | access to parameter i | negative strictlyNegative |
| SignAnalysis.cs:65:21:65:21 | access to parameter j | negative strictlyNegative |
| SignAnalysis.cs:67:13:67:21 | ... = ... | negative |
| SignAnalysis.cs:67:17:67:17 | access to parameter i | negative strictlyNegative |
| SignAnalysis.cs:67:17:67:21 | ... % ... | negative |
| SignAnalysis.cs:67:21:67:21 | access to parameter j | negative strictlyNegative |
| SignAnalysis.cs:68:38:68:38 | access to local variable x | negative |
| SignAnalysis.cs:69:13:69:19 | ... = ... | negative strictlyNegative |
| SignAnalysis.cs:69:17:69:17 | access to parameter i | negative strictlyNegative |
| SignAnalysis.cs:69:17:69:19 | ...++ | negative strictlyNegative |
| SignAnalysis.cs:70:38:70:38 | access to local variable x | negative strictlyNegative |
| SignAnalysis.cs:71:13:71:19 | ... = ... | negative |
| SignAnalysis.cs:71:17:71:17 | access to parameter i | negative |
| SignAnalysis.cs:71:17:71:19 | ...-- | negative |
| SignAnalysis.cs:72:38:72:38 | access to local variable x | negative |
| SignAnalysis.cs:73:13:73:18 | ... = ... | positive strictlyPositive |
| SignAnalysis.cs:73:17:73:18 | -... | positive strictlyPositive |
| SignAnalysis.cs:73:18:73:18 | access to parameter i | negative strictlyNegative |
| SignAnalysis.cs:74:38:74:38 | access to local variable x | positive strictlyPositive |
| SignAnalysis.cs:75:13:75:18 | ... = ... | negative strictlyNegative |
| SignAnalysis.cs:75:17:75:18 | +... | negative strictlyNegative |
| SignAnalysis.cs:75:18:75:18 | access to parameter i | negative strictlyNegative |
| SignAnalysis.cs:76:38:76:38 | access to local variable x | negative strictlyNegative |
| SignAnalysis.cs:77:21:77:27 | (...) ... | negative strictlyNegative |
| SignAnalysis.cs:77:27:77:27 | access to parameter i | negative strictlyNegative |
| SignAnalysis.cs:78:38:78:38 | access to local variable l | negative strictlyNegative |
| SignAnalysis.cs:80:13:80:17 | ... = ... | negative strictlyNegative |
| SignAnalysis.cs:80:17:80:17 | access to parameter i | negative strictlyNegative |
| SignAnalysis.cs:81:13:81:13 | access to local variable x | negative strictlyNegative |
| SignAnalysis.cs:81:13:81:18 | ... + ... | negative strictlyNegative |
| SignAnalysis.cs:81:13:81:18 | ... = ... | negative strictlyNegative |
| SignAnalysis.cs:81:18:81:18 | access to parameter i | negative strictlyNegative |
| SignAnalysis.cs:82:38:82:38 | access to local variable x | negative strictlyNegative |
| SignAnalysis.cs:87:21:87:21 | access to parameter i | negative strictlyNegative |
| SignAnalysis.cs:87:25:87:25 | access to parameter j | positive strictlyPositive |
| SignAnalysis.cs:89:13:89:21 | ... = ... | negative strictlyNegative |
| SignAnalysis.cs:89:17:89:17 | access to parameter i | negative strictlyNegative |
| SignAnalysis.cs:89:17:89:21 | ... * ... | negative strictlyNegative |
| SignAnalysis.cs:89:21:89:21 | access to parameter j | positive strictlyPositive |
| SignAnalysis.cs:90:38:90:38 | access to local variable x | negative strictlyNegative |
| SignAnalysis.cs:91:13:91:21 | ... = ... | negative |
| SignAnalysis.cs:91:17:91:17 | access to parameter i | negative strictlyNegative |
| SignAnalysis.cs:91:17:91:21 | ... / ... | negative |
| SignAnalysis.cs:91:21:91:21 | access to parameter j | positive strictlyPositive |
| SignAnalysis.cs:92:38:92:38 | access to local variable x | negative |
| SignAnalysis.cs:93:13:93:21 | ... = ... | negative strictlyNegative |
| SignAnalysis.cs:93:17:93:17 | access to parameter i | negative strictlyNegative |
| SignAnalysis.cs:93:17:93:21 | ... - ... | negative strictlyNegative |
| SignAnalysis.cs:93:21:93:21 | access to parameter j | positive strictlyPositive |
| SignAnalysis.cs:94:38:94:38 | access to local variable x | negative strictlyNegative |
| SignAnalysis.cs:95:13:95:21 | ... = ... | negative |
| SignAnalysis.cs:95:17:95:17 | access to parameter i | negative strictlyNegative |
| SignAnalysis.cs:95:17:95:21 | ... % ... | negative |
| SignAnalysis.cs:95:21:95:21 | access to parameter j | positive strictlyPositive |
| SignAnalysis.cs:96:38:96:38 | access to local variable x | negative |
| SignAnalysis.cs:97:21:97:21 | access to parameter i | negative strictlyNegative |
| SignAnalysis.cs:97:25:97:25 | access to parameter j | positive strictlyPositive |
| SignAnalysis.cs:104:17:104:20 | 4.2 | positive strictlyPositive |
| SignAnalysis.cs:105:34:105:34 | access to local variable f | positive strictlyPositive |
| SignAnalysis.cs:106:17:106:19 | 4.2 | positive strictlyPositive |
| SignAnalysis.cs:107:34:107:34 | access to local variable d | positive strictlyPositive |
| SignAnalysis.cs:108:18:108:21 | 4.2 | positive strictlyPositive |
| SignAnalysis.cs:109:34:109:35 | access to local variable de | positive strictlyPositive |
| SignAnalysis.cs:110:17:110:19 | a | positive strictlyPositive |
| SignAnalysis.cs:111:34:111:34 | access to local variable c | positive strictlyPositive |
| SignAnalysis.cs:120:9:120:10 | access to field f0 | positive |
| SignAnalysis.cs:120:9:120:12 | ...++ | positive |
| SignAnalysis.cs:121:34:121:35 | access to field f0 | positive strictlyPositive |
| SignAnalysis.cs:122:9:122:10 | access to field f0 | positive |
| SignAnalysis.cs:129:9:129:16 | ... = ... | negative strictlyNegative |
| SignAnalysis.cs:129:14:129:16 | -... | negative strictlyNegative |
| SignAnalysis.cs:129:15:129:16 | 10 | positive strictlyPositive |
| SignAnalysis.cs:141:34:141:34 | access to local variable i | positive strictlyPositive |
| SignAnalysis.cs:148:20:148:20 | access to parameter x | negative strictlyNegative |
| SignAnalysis.cs:151:18:151:18 | 1 | positive strictlyPositive |
| SignAnalysis.cs:153:20:153:20 | access to parameter y | positive strictlyPositive |
| SignAnalysis.cs:156:18:156:19 | -... | negative strictlyNegative |
| SignAnalysis.cs:156:19:156:19 | 1 | positive strictlyPositive |
| SignAnalysis.cs:161:13:161:13 | access to parameter x | positive |
| SignAnalysis.cs:163:20:163:20 | access to parameter y | positive strictlyPositive |
| SignAnalysis.cs:166:22:166:22 | 1 | positive strictlyPositive |
| SignAnalysis.cs:169:20:169:20 | access to parameter y | positive strictlyPositive |
| SignAnalysis.cs:177:17:177:17 | 1 | positive strictlyPositive |
| SignAnalysis.cs:178:13:178:13 | access to local variable i | positive strictlyPositive |
| SignAnalysis.cs:180:38:180:38 | access to local variable i | positive strictlyPositive |
| SignAnalysis.cs:186:34:186:44 | access to property Length | positive |
| SignAnalysis.cs:187:16:187:36 | 3 | positive strictlyPositive |
| SignAnalysis.cs:187:28:187:28 | 1 | positive strictlyPositive |
| SignAnalysis.cs:187:31:187:31 | 2 | positive strictlyPositive |
| SignAnalysis.cs:187:34:187:34 | 3 | positive strictlyPositive |
| SignAnalysis.cs:188:34:188:44 | access to property Length | positive strictlyPositive |
| SignAnalysis.cs:189:34:189:45 | call to method Count | positive |
| SignAnalysis.cs:190:34:190:55 | call to method Count | positive |
| SignAnalysis.cs:190:54:190:54 | 1 | positive strictlyPositive |
| SignAnalysis.cs:193:34:193:41 | access to property Length | positive |
| SignAnalysis.cs:196:34:196:51 | call to method Count | positive |
| SignAnalysis.cs:198:17:198:59 | 2 | positive strictlyPositive |
| SignAnalysis.cs:198:17:198:59 | 3 | positive strictlyPositive |
| SignAnalysis.cs:198:32:198:32 | 1 | positive strictlyPositive |
| SignAnalysis.cs:198:35:198:35 | 1 | positive strictlyPositive |
| SignAnalysis.cs:198:42:198:42 | 1 | positive strictlyPositive |
| SignAnalysis.cs:198:45:198:45 | 2 | positive strictlyPositive |
| SignAnalysis.cs:198:52:198:52 | 1 | positive strictlyPositive |
| SignAnalysis.cs:198:55:198:55 | 3 | positive strictlyPositive |
| SignAnalysis.cs:199:34:199:41 | access to property Length | positive strictlyPositive |
| SignAnalysis.cs:206:38:206:38 | access to parameter i | positive strictlyPositive |
| SignAnalysis.cs:210:38:210:38 | access to parameter i | negative |
| SignAnalysis.cs:219:38:219:38 | access to parameter i | positive strictlyPositive |
| SignAnalysis.cs:223:17:223:17 | access to parameter i | negative |
| SignAnalysis.cs:225:42:225:42 | access to parameter i | negative strictlyNegative |
| SignAnalysis.cs:236:38:236:38 | access to parameter i | positive strictlyPositive |
| SignAnalysis.cs:240:17:240:17 | access to parameter i | negative |
| SignAnalysis.cs:242:42:242:42 | access to parameter i | negative strictlyNegative |
| SignAnalysis.cs:257:17:257:17 | access to parameter i | positive |
| SignAnalysis.cs:257:17:257:19 | ...-- | positive |
| SignAnalysis.cs:260:38:260:38 | access to parameter i | negative strictlyNegative |
| SignAnalysis.cs:267:17:267:17 | access to parameter j | positive strictlyPositive |
| SignAnalysis.cs:267:17:267:19 | ...-- | positive strictlyPositive |
| SignAnalysis.cs:268:42:268:42 | access to parameter j | positive |
| SignAnalysis.cs:270:38:270:38 | access to parameter j | negative |
| SignAnalysis.cs:277:17:277:17 | access to parameter k | positive strictlyPositive |
| SignAnalysis.cs:277:17:277:19 | ...-- | positive strictlyPositive |
| SignAnalysis.cs:278:42:278:42 | access to parameter k | positive |
| SignAnalysis.cs:280:21:280:21 | access to parameter k | positive |
| SignAnalysis.cs:280:26:280:26 | 5 | positive strictlyPositive |
| SignAnalysis.cs:291:22:291:22 | access to parameter i | positive strictlyPositive |
| SignAnalysis.cs:292:34:292:34 | access to parameter i | positive strictlyPositive |
| SignAnalysis.cs:295:38:295:38 | access to parameter i | positive strictlyPositive |
